MIP-58 Simplified Bridge Design (#58)
movementlabsxyz/MIP: MIP/mip-58/README.md

# MIP-28: Simplified Bridge Design

- **Description**: Proposes a simplified bridge design to address inefficiencies, security risks, and high costs in the current implementation.

> - **Authors**: [Author](mailto:primata@movementlabs.xyz)

Reviewer: ```suggestion
- **Authors**: [Primata](mailto:primata@movementlabs.xyz)
```

3. **Unfriendly User Experience**:
   - The current design requires users to have funds on the destination chain for finalization. This increases friction and discourages adoption.
   - Sponsored transactions are essential but remain unimplemented, leaving users stranded without funds.
4. **Security Risks**:
   - Relayer keys are a critical point of failure.
   - Refund logic introduces vulnerabilities where attackers could exploit the rate limit and relayer downtime.
   - For instance, if the relayer fails to finalize on the initiator, a malicious actor could:
     - Take over the refund keys.
     - Exploit both directions of the bridge (e.g., ETH→MOVE and MOVE→ETH) repeatedly, draining funds.
5. **Maintenance Burden**:
   - The current audit is outdated and does not reflect the significant changes made since.
   - The current design requires a complete UI/UX overhaul, adding complexity and delay.
6. **Unnecessary Complexity**:
   - HTLC-based bridges are largely abandoned in favor of simpler, more effective designs.

>    - Examples like the Consensys HTLC bridge demonstrate the pitfalls of such approaches.

Reviewer: Here's a link to an article by Consensys citing Connext’s Amarok upgrade as an example of moving away from a so-called atomic swap (HTLC) design: https://entethalliance.org/crosschain-bridges-overview-where-we-are-now/

You could also link to the old Consensys HTLC bridge code https://github.com/Consensys/htlc-bridge to show that it is no longer in development in favor of another model.

7. **Infrastructure Simplification**:

>    - Infrastructure is still incomplete for the current HTLC bridge design and we are struggling to find sound solutions that balance UI/UX and security.

Reviewer: ```suggestion
   - Infrastructure is still incomplete for the current HTLC bridge design and we are struggling to find sound solutions that balance UI/UX and security. For example, [this issue](https://github.com/movementlabsxyz/movement/issues/838) regarding UX vs number of confirmations to require on the Ethereum side, is still under debate.
```

7. **Batch Processing for Downtime**:
   - Multi-signature relayers can process multiple pending transactions in a single batch during downtime.

8. **BridgeTransferId**:
   - Continue using unique identifier to prevent double-spending and track transactions securely.

9. **Best Practices**:
   - Adopt currently used bridge designs from established designs like Arbitrum, LayerZero and Blast bridges which use a relayer to finalize the bridge.
   - User is not required to have funds on counterparty contract to finalize the bridge.

## Exploits and Potential Losses

1. **Rate-Limit Exploits**:
   - If the relayer is down, an attacker could exploit refund functionality to repeatedly bridge funds within the rate limit.
2. **Key Compromise**:
   - Relayer or refund keys could be compromised, leading to unauthorized transactions.
3. **Fee Misestimation**:
   - Incorrect fee calculations (e.g., underestimating gas) can cause significant financial losses for the protocol.
4. **User Abandonment**:
   - High costs and complexity deter users, leading to lower adoption and reputational damage.

## Reference Implementation

1. **Two-Transaction Flow**:
   - User initiates the transfer.

>    - Relayer or multisig completes the transfer with parameter validation.

Reviewer: Could you add something about how users are supposed to be able to check on-chain that their transfer is complete? Does the relayer keep some database where users can look up transactions associated with a specific bridge transfer ID?

2. **Batch Completion**:
   - Multisig relayers process pending transactions in batches during downtime, ensuring timely resolution.

3. **Contract Simplification**:
   - Combine lock and completion functionality on the counterparty contract.
   - Remove refund logic to streamline operations and improve security.

> 4. **Gas Optimization**:

Reviewer: Minimizing gas usage as a benefit is already covered in `5. **Cost Efficiency**` above. 

Here I think it would be good to add the code for the respective functions on the Solidity and Move sides.

Author: Should it be added here?

Reviewer: Maybe it could be a good place, yes, like under each bullet point:

```
 - Combine lock and completion functionality on the counterparty contract.
 - Remove refund logic to streamline operations and improve security.
```
You could add some code and/or a diagram.

Reviewer 2: after conversation with @andygolay this is resolved.

## Specification

The simplified bridge design focuses on minimizing complexity and maximizing security:

### Protocol Description

The bridge is a simpler version of the current implementation, which relies on the assumption that the Relayer Keys are not to be compromised.

L1 -> L2

1. User initiates a bridge transfer on L1. Contract stores a mapping of the user bridgeTransferIds for easy access. It transfers from the user MOVE amount to the contract. Transaction emits originator, recipient, amount and nonce.
2. Relayer awaits for finalization of the transaction on L1.
3. Relayer completes the bridge on L2.
4. Completion transaction verifies that the transaction is truthful by comparing the provided bridgeTransferId hash and the emitted values of originator, recipient, amount and nonce. Finally, it mints MOVE to the recipient address.
5. User is notified on the frontend that their transaction has been completed.
6. In case of a rare case of a reorg on the recipient network that leads to the transaction removal from the user perspective, the transaction can be redone and relayer multisig can handle it.

![L1-L2](l1l2.png)

L2 -> L1

> 1. User initiates a bridge transfer on L2. Contract stores a mapping of the user bridgeTransferIds for easy access. It transfers from the user MOVE amount to the contract and burns it. Transaction emits originator, recipient, amount and nonce.

Reviewer: Actually it's a mapping of the nonces to bridge transfer details, right?

Reviewer 3: yes, I will be used by the relayer to bootstrap.

Reviewer 2: @Primata has this been adressed?

2. Relayer awaits for finalization of the transaction on L2.
3. Relayer completes the bridge on L1.
4. Completion transaction verifies that the transaction is truthful by comparing the provided bridgeTransferId hash and the emitted values of originator, recipient, amount and nonce. Finally, it transfers MOVE to the recipient address, which has been previously bridged from L1.
5. User is notified on the frontend that their transaction has been completed.
6. In case of a rare caase of a reorg on the recipient network that leads to the transaction removal from the user perspective, the transaction can be redone and relayer multisig can handle it.

![L2-L1](l2l1.png)

### Key Features

> 1. **Two-Transaction Model**:

Reviewer 2: ```suggestion
1. **Lock/Mint mechanism**:
```

     - **Automated Relayer**: Operated with minimal human involvement; its private key is highly secured.
     - **Multi-Signature Relayer**: Managed by the team to guarantee transaction completion in case of failures.

5. **Cost Efficiency**:
   - Minimize gas costs by reducing the number of interactions and simplifying fee calculations.

6. **Enhanced Security**:
   - Avoid refund logic to close exploit windows.
   - Protect against key compromise through key isolation or known by no parties and multi-signature relayer setups.
   - There is no scenario where a bridge could lead to double-spending. It's either completed by relayers or not.
   - User currently can loose its bridge `preImage` which could lead to them being unable to complete the bridge. By not relying on a `preImage` from the user, it minimizes issues. It is not a loss in security because the purpose of the `preImage` is solely for refunding.

7. **Batch Processing for Downtime**:
   - Multi-signature relayers can process multiple pending transactions in a single batch during downtime.

8. **BridgeTransferId**:
   - Continue using unique identifier to prevent double-spending and track transactions securely.

> 9. **Bridge Fee**:

Reviewer 4: https://github.com/movementlabsxyz/movement/issues/781

Reviewer 4: @apenzk @franck44 wondering if you feel like this is a valid approach.

Reviewer 2: responded at the link you shared

   - Because of the over-engineered design, infrastructure is prone to error and we might end up being damaged by the amount of infrastructure we have not built yet and has to be built for the relayer to fully function.
   - We could strip down the Relayer code and achieve a final design much more quickly.

The proposed lock/mint bridge design mitigates these issues, creating a safer, faster, and user-friendly bridge while maintaining operational reliability.

## Specification

The lock/mint bridge design focuses on minimizing complexity and maximizing security:

### Protocol Description

The bridge is a simpler version of the current implementation, which relies on the assumption that the Relayer Keys are not to be compromised.

L1 -> L2

1. User initiates a bridge transfer on L1. Contract stores a mapping of the user bridgeTransferIds for easy access. It transfers from the user MOVE amount to the contract. Transaction emits originator, recipient, amount and nonce.
2. Relayer awaits for finalization of the transaction on L1.
3. Relayer completes the bridge on L2.

> 4. Completion transaction verifies that the transaction is truthful by comparing the provided bridgeTransferId hash and the emitted values of originator, recipient, amount and nonce. Finally, it mints MOVE to the recipient address.

Reviewer: The method for checking the parameters needs to be specified. Where are the code samples we put in? Those should be restored here and the L2 -> L1 steps (for each, include initiate function code in step 1 and complete function code in step 4). Right now these steps are unacceptably vague; it doesn't specify how the comparison happens. Linking to a couple GitHub repositories later in the document is not sufficient. The code should be here in the protocol description, or at least there should be an exact description of how the comparison happens.

      1. Relayer sets a fee on L2, a global variable that can be set at any given time. Considering that block validation time on L1 is bigger than on L2, it becomes a viable approach since L2 can rapidly adjust the fee according to the current block and always charge an above gas cost fee to assure that despite hiccups, the bridge is net positive. $MOVE is deducted from the amount of tokens that are currently being bridged and transferred to a funds manager. Input of the user is on the final value of tokens that it should receive on the L1. This gives the protocol a very reliable way to estimate how much MOVE will be charged and feed to the user a precise amount of tokens.
      2. Enable the Relayer to specify on the L1 completeBridgeTransfer transaction, the bridge fee per transaction. The amount is deducted from the total amount of tokens that were bridged and transferred to a funds manager. The dangerous situation is that we expect to way for over 10 minutes before the transfer can occur, and this could lead to a big disparity between the expected amout of funds and the actual amount of tokens received.

> [Solidity Implementation](https://github.com/movementlabsxyz/movement/tree/primata/simple-native-bridge)

Reviewer: I think the code samples should be restored. In particular the code for initiate and complete, in both Move and Solidity, in particular showing how the parameter check works.

Reviewer 4: but this is the code itself, isn't it?

        // Generate a unique nonce to prevent replay attacks, and generate a transfer ID
        bridgeTransferId =
            keccak256(abi.encodePacked(originator, recipient, amount, ++_nonce));

        // We have all bridgeTransferIds available by user because we don't have to re-access it.
        initiatedBridgeTransfers[originator].add(bridgeTransferId); 

        emit BridgeTransferInitiated(bridgeTransferId, originator, recipient, amount, _nonce);
        return bridgeTransferId;
    }

function completeBridgeTransfer(
        bytes32 bridgeTransferId,
        bytes32 originator,
        address recipient,
        uint256 amount,
        uint256 nonce
        ) external onlyRole(RELAYER_ROLE) {
         require(bridgeTransferId == keccak256(abi.encodePacked(originator, recipient, amount, nonce)), InvalidBridgeTransferId());
        require(!completedBridgeTransfers[bridgeTranserId]);
        if (moveToken.transfer(recipient, amount)) revert MOVETransferFailed();

        completedBridgeTransfers[bridgeTranserId] = true;

        emit BridgeTransferCompleted(bridgeTransferId, originator, recipient, amount, nonce);
    }
```

```move

> // tbd

Reviewer: ```suggestion

    /// Module-level resource to store the nonce
    struct Nonce has key {
        value: u64,
    }

    /// Creates bridge transfer details with validation.
    ///
    /// @param initiator The initiating party of the transfer.
    /// @param recipient The receiving party of the transfer.
    /// @param amount The amount to be transferred.
    /// @param nonce The unique nonce for the transfer.
    /// @return A `BridgeTransferDetails` object.
    /// @abort If the amount is zero.
    public(friend) fun create_details<Initiator: store, Recipient: store>(
        initiator: Initiator,
        recipient: Recipient,
        amount: u64,
        nonce: u64
    ): BridgeTransferDetails<Initiator, Recipient> {
        assert!(amount > 0, EZERO_AMOUNT);

        BridgeTransferDetails {
            addresses: AddressPair {
                initiator,
                recipient
            },
            amount,
            nonce,
            state: PENDING_TRANSACTION,
        }
    }
    
    /// Increment and get the current nonce
    fun increment_and_get_nonce(signer: address): u64 {
        let nonce_ref = borrow_global_mut<Nonce>(signer);
        nonce_ref.value = nonce_ref.value + 1;
        nonce_ref.value
    }
    
    
        
/// Initiate a bridge transfer of MOVE from Movement to the base layer
/// Anyone can initiate a bridge transfer from the source chain
/// The amount is burnt from the initiator
public entry fun initiate_bridge_transfer(
    initiator: &signer,
    recipient: vector<u8>,
    amount: u64
) {
    let initiator_address = signer::address_of(initiator);
    let ethereum_address = ethereum::ethereum_address(recipient);

    // Increment and retrieve the nonce
    let nonce = increment_and_get_nonce(initiator_address);

    // Create bridge transfer details
    let details = atomic_bridge_store::create_details(
        initiator_address,
        ethereum_address, 
        amount,
        nonce
    );

    // Generate a unique bridge transfer ID
    let bridge_transfer_id = bridge_transfer_id(&details);

    // Add the transfer details to storage
    atomic_bridge_store::add(bridge_transfer_id, details);

    // Burn the amount from the initiator
    atomic_bridge::burn(initiator_address, amount);

    // Emit an event with nonce
    event::emit(
        BridgeTransferInitiatedEvent {
            bridge_transfer_id,
            initiator: initiator_address,
            recipient,
            amount,
            nonce,
        }
    );
}
```

```
    /// Completes a bridge transfer by the initiator.
    ///
    /// @param caller The signer representing the bridge operator.
    /// @param initiator The initiator's Ethereum address as a vector of bytes.
    /// @param bridge_transfer_id The unique identifier for the bridge transfer.
    /// @param hash_lock The hash lock for securing the transfer.
    /// @param time_lock The time lock duration for the transfer.
    /// @param recipient The address of the recipient on the Aptos blockchain.
    /// @param amount The amount of assets to be locked.
    /// @abort If the caller is not the bridge operator.
    public entry fun complete_bridge_transfer(
        caller: &signer,
        initiator: vector<u8>,
        bridge_transfer_id: vector<u8>,
        hash_lock: vector<u8>,
        recipient: address,
        amount: u64,
        nonce: u64
    ) {
        atomic_bridge_configuration::assert_is_caller_operator(caller);
        let ethereum_address = ethereum::ethereum_address(initiator);
        let time_lock = atomic_bridge_store::create_time_lock(
            atomic_bridge_configuration::counterparty_timelock_duration());
        let details = atomic_bridge_store::create_details(
            ethereum_address,
            recipient,
            amount,
           nonce
        );

        // bridge_store::add_counterparty(bridge_transfer_id, details);
        atomic_bridge_store::add(bridge_transfer_id, details);

        event::emit(
            BridgeTransferCompletedEvent {
                bridge_transfer_id,
                initiator,
                recipient,
                amount,
                nonce,
            },
        );
    }
```

   ```
   entry fun set_fee(caller: &signer, fee: u64) {
      assert_is_maintainer(caller);
      borrow_global<BridgeConfig>(@aptos_framework).fee = fee;
   }
   ```

10. **Best practices**

- Adopt currently used bridge designs from established designs like Arbitrum, LayerZero and Blast bridges, which use a Relayer to finalize the bridge.
- User is not required to have funds on counterparty contract to finalize the bridge.

## Exploits and potential losses

1. **Key compromise**
   - The compromise of the Relayer keys would lead to unauthorized transactions, putting the protocol at risk of unlimited value exploit.  The protocol must absorb the losses and rotate Relayers. To minimize this risk [MIP-74](https://github.com/movementlabsxyz/MIP/pull/74) proposes a rate limitation even for the Relayer.
2. **Fee wrong estimation**
   - Incorrect fee calculations (e.g., underestimating gas) can cause significant financial losses for the operator.

## Reference implementation

> [Solidity Implementation](https://github.com/movementlabsxyz/movement/tree/primata/simple-native-bridge)

Reviewer 2: as per discussion with @andygolay we should reference specific commits. Could you recommend a specific commit? 

also for the Move Implementation

## Motivation

The current bridge design poses numerous challenges, including inefficiency, cost, and user frustration:

1. **Transaction Complexity**: Requires four interactions:
   - Two by the user (initiation and finalization).
   - One by the relayer to complete the transaction.
   - One by a third party (tasked by the relayer) to finalize refunds in case of failure.
   - If any of these fails, users and the protocol face losses.
2. **High Cost**:
   - The multi-transaction setup is expensive for both users and the protocol.
   - Misestimations in fee calculations could result in significant losses.
   - Example: At 16 gwei, bridging costs $20 for 10k transactions. If fees are miscalculated by 25%, this leads to a $25k loss. At 100 gwei, the loss escalates dramatically.
3. **Unfriendly User Experience**:
   - The current design requires users to have funds on the destination chain for finalization. This increases friction and discourages adoption.
   - Sponsored transactions are essential to the current implementation but remain unimplemented, leaving users stranded without funds.
4. **Security Risks**:

>    - Relayer keys are a critical point of failure. This assumption cannot be avoided.

Author: What I meant here is that we keep the same assumption, but this design has less assumptions overall. E.g. If the relayer goes down, we are not at risk.

>    - Refund logic introduces vulnerabilities where attackers could exploit the rate limit and relayer downtime.

Author: Well, I mean, it has to be specified. Rate limiter is a sound solution and we should definitely keep it, but there is a possibility of refunding ocurring anyways if complete on counterparty doesn't go live in time for example.

   - For instance, if the relayer fails to finalize on the initiator, a malicious actor could:

>      - Take over the refund keys.

Author: By delegating to a multisig, someone could:
have the keys of refunders, bridge back and forth several times, and refund from both sides using refunder keys. Both proposals have multisigs and we are separating them as a concern. Because the relayer multisig on this design is only completing, the rate limit would limit the damage. Refunding is not covered by the rate limiter and I don't thing we can design around that besides by having the observer.

     - Exploit both directions of the bridge (e.g., ETH→MOVE and MOVE→ETH) repeatedly, draining funds.
5. **Maintenance Burden**:

>    - The current audit is outdated and does not reflect the significant changes made since.

Reviewer 2: ```suggestion
```

>    - The current audit is outdated and does not reflect the significant changes made since.

Reviewer 4: I still think the audit is outdated, we have included sponsored txs and rate limiting.

   - The current design requires a complete UI/UX overhaul, adding complexity and delay.
6. **Unnecessary Complexity**:
   - HTLC-based bridges are largely abandoned in favor of simpler, more effective designs.

>    - Examples like the [Consensys HTLC](https://github.com/Consensys/htlc-bridge) bridge demonstrate the [pitfalls of such approaches](https://entethalliance.org/crosschain-bridges-overview-where-we-are-now/).

Author: yes, but it still has to be indicated. I don't think a lot of people knew what it entitled as we were developing it.

>    - Examples like the [Consensys HTLC](https://github.com/Consensys/htlc-bridge) bridge demonstrate the [pitfalls of such approaches](https://entethalliance.org/crosschain-bridges-overview-where-we-are-now/).

Reviewer 2: As per committee agreement, merging will go ahead without addressing this comment. Please consider to raise an issue if this is of concern @Primata

7. **Infrastructure Simplification**:
   - Infrastructure is still incomplete for the current HTLC bridge design and we are struggling to find sound solutions that balance UI/UX and security. For example, [this issue](https://github.com/movementlabsxyz/movement/issues/838) regarding UX vs number of confirmations to require on the Ethereum side, is still under debate.

>    - Because of the over-engineered design, infrastructure is prone to error and we might end up being damaged by the amount of infrastructure we have not built yet and has to be built for the relayer to fully function.

Author: I would say it's over-engineered for the purpose of a bridge. The point is RFC-40 design proposal lead us to a point where we will be damaging users.

Sponsored transactions is one example. And polishing the relayer and indexer that contains 8 events to handle instead of 4 and has to keep track of all the intermediary calls. In the meantime we are wondering if we should include or not an essential finality await because it's a big issue design wise. The other option the user just starts a bridge and receives the funds on the specified address and we reduce the stack, meat the requirements for a bridge with less assumptions and make it safer by default.

   - We could strip down the Relayer code and achieve a final design much more quickly.

The proposed two-transaction design mitigates these issues, creating a safer, faster, and user-friendly bridge while maintaining operational reliability.

## Specification

The simplified bridge design focuses on minimizing complexity and maximizing security:

### Protocol Description

The bridge is a simpler version of the current implementation, which relies on the assumption that the Relayer Keys are not to be compromised.

L1 -> L2

1. User initiates a bridge transfer on L1. Contract stores a mapping of the user bridgeTransferIds for easy access. It transfers from the user MOVE amount to the contract. Transaction emits originator, recipient, amount and nonce.
2. Relayer awaits for finalization of the transaction on L1.
3. Relayer completes the bridge on L2.

> 4. Completion transaction verifies that the transaction is truthful by comparing the provided bridgeTransferId hash and the emitted values of originator, recipient, amount and nonce. Finally, it mints MOVE to the recipient address.

Reviewer 5: My main question here is, is this safe? I would need to dig into this, which I don't have the time now due to our current obligations.

Author: It's the current implementation stripped down to the relayer making the transfer directly and we remoivng the refund capability which produces more risk then having a relayer that is capable of completing bridge transactions, which is subject to rate limiting.

> # MIP-58: Simplified Bridge Design

Reviewer 2: ```suggestion
# MIP-58: Trusted-Relayer focused Bridge Design
```

> - **Description**: Proposes a simplified bridge design to address inefficiencies, security risks, and high costs in the current implementation.

Reviewer 2: ```suggestion
- **Description**: Proposes a bridge design that capitalizes on the trust-assumption on the relayer, reduces the number of messages, and potentially cost.
```

   - Example: At 16 gwei, bridging costs $20 for 10k transactions. If fees are miscalculated by 25%, this leads to a $25k loss. At 100 gwei, the loss escalates dramatically.
3. **Unfriendly User Experience**:
   - The current design requires users to have funds on the destination chain for finalization. This increases friction and discourages adoption.
   - Sponsored transactions are essential to the current implementation but remain unimplemented, leaving users stranded without funds.
4. **Security Risks**:
   - Relayer keys are a critical point of failure. This assumption cannot be avoided.
   - Refund logic introduces vulnerabilities where attackers could exploit the rate limit and relayer downtime.
   - For instance, if the relayer fails to finalize on the initiator, a malicious actor could:
     - Take over the refund keys.
     - Exploit both directions of the bridge (e.g., ETH→MOVE and MOVE→ETH) repeatedly, draining funds.
5. **Maintenance Burden**:
   - The current audit is outdated and does not reflect the significant changes made since.
   - The current design requires a complete UI/UX overhaul, adding complexity and delay.
6. **Unnecessary Complexity**:

>    - HTLC-based bridges are largely abandoned in favor of simpler, more effective designs.

Reviewer 6: HTLC is used for _atomic swaps_. [Chainlink](https://chain.link/education-hub/atomic-swaps) has an explainer for atomic swaps. 
HTLC-bridges are used to implement atomic swaps between parties that do not trust each other. 
HTLC is not _largely_ abandoned (see the link above). 
The problem is that we have used an atomic swap architecture [RFC-40](https://github.com/movementlabsxyz/rfcs/blob/main/0040-atomic-bridge/rfc-0040-atomic-bridge.md) to guide the implementation of the bridge. 
But a bridge (transfer) is different to a swap: in a swap parties can withdraw from the deal before the end of an agreed period. They also have to escrow funds.
In a bridge, one party wants to _transfer_ assets from one chain A to another chain B.
Most bridges implement a lock/mint or mint/burn mechanism, not an atomic swap.

Reviewer 2: As per committee agreement, merging will go ahead without addressing this comment. Please consider to raise an issue if this is of concern @franck44

>    - Examples like the [Consensys HTLC](https://github.com/Consensys/htlc-bridge) bridge demonstrate the [pitfalls of such approaches](https://entethalliance.org/crosschain-bridges-overview-where-we-are-now/).

Reviewer 6: I don't see how the first link demonstrates the pitfalls of such approaches. 
The first link is outdated (in fact it was not used by Consensys, but rather GPACT was considered, and then abandoned).
The second link [Crosschains, where are we now](https://entethalliance.org/crosschain-bridges-overview-where-we-are-now/) is an overview of the cross chain solutions and their main features. It does not demonstrate the pitfalls of the [Consensys HTLC bridge](https://github.com/Consensys/htlc-bridge).

> This method is trust minimized, but requires both parties to stay online for the duration of **the swap** in order to withdraw funds on the other side, which can cause friction for end-users.
> It seems that each design has some pro and some con.

This applies to an **atomic swap**. It is unclear to me why there was a decision to use a swap architecture to implement a bridge. I mentioned it 3 weeks ago [here](https://github.com/movementlabsxyz/MIP/pull/39#issuecomment-2448794875).

Reviewer 2: As per committee agreement, merging will go ahead without addressing this comment. Please consider to raise an issue if this is of concern @franck44

7. **Infrastructure Simplification**:

>    - Infrastructure is still incomplete for the current HTLC bridge design and we are struggling to find sound solutions that balance UI/UX and security. For example, [this issue](https://github.com/movementlabsxyz/movement/issues/838) regarding UX vs number of confirmations to require on the Ethereum side, is still under debate.

Reviewer 6: This issue does not relate to HTLC. This issue is not going to be resolved by a new architecture. In a cross chain "deal" (bridge, atomic swap etc) the two chains need to communicate their states to the other chain. The communicated states must be trusted by each party. 
In a trustless set up, the two chains may run a (light)-client of the other chain to verify the "validity" of a block/state (using Merkle proofs etc).
In a trusted set up, e.g., the relayer is trusted, the relayer is trusted to relay information faithfully (covering events to transactions). But we have to make sure that the information about the state is trustworthy i.e. that the states that are communicated (relayed) by the relayer are final or not going to be reverted (re-orgd) .

In the current state of Ethereum we know that:
1. re-org occurs approximately once every hour
2. no-reorg on finalised blocks.

In between (k-confirmations) we don't have data or evidence that it is safe to use the blocks/states. 
If we choose this route, we have to take into account possible re-orgs which arguably complicates the bridge design.

   - Because of the over-engineered design, infrastructure is prone to error and we might end up being damaged by the amount of infrastructure we have not built yet and has to be built for the relayer to fully function.
   - We could strip down the Relayer code and achieve a final design much more quickly.

> The proposed two-transaction design mitigates these issues, creating a safer, faster, and user-friendly bridge while maintaining operational reliability.

Reviewer 6: This proposal is the implementation of a lock/mint bridge. 
We may reference some known explainers like [ChainLink Cross-chain bridge](https://chain.link/education-hub/cross-chain-bridge)?

## Specification

> The simplified bridge design focuses on minimizing complexity and maximizing security:

Reviewer 6: It is not a simplified bridge design but a lock/mint design.
So it is not new.

Reviewer 2: i agree lets refrain from "simplified bridge". 

if this is a lock/mint.. how do we call the previous version (the HTLC-based bridge). would it be a swap-type bridge? (this is relevant for the [MIP-60: crosschain-bridge-design](https://github.com/movementlabsxyz/MIP/pull/60))

Reviewer 6: The "HTLC -bridge" is using a swap mechanism to implement a bridge (which is not optimal).
I think I have tried to explain the differences between (atomic) swap and bridge in (MIP-60)[https://github.com/movementlabsxyz/MIP/pull/60).

>  swap-type bridge?

It is more about treating a "bridge" as "swap" but that's overkill. It is related to the initial proposal [RFC-40](https://github.com/movementlabsxyz/rfcs/blob/main/0040-atomic-bridge/rfc-0040-atomic-bridge.md) which describes an atomic swap and was used as a specification for the bridge..

5. User is notified on the frontend that their transaction has been completed.
6. In case of a rare case of a reorg on the recipient network that leads to the transaction removal from the user perspective, the transaction can be redone and relayer multisig can handle it.

![L1-L2](l1l2.png)

L2 -> L1

1. User initiates a bridge transfer on L2. Contract stores a mapping of the user bridgeTransferIds for easy access. It transfers from the user MOVE amount to the contract and burns it. Transaction emits originator, recipient, amount and nonce.
2. Relayer awaits for finalization of the transaction on L2.
3. Relayer completes the bridge on L1.
4. Completion transaction verifies that the transaction is truthful by comparing the provided bridgeTransferId hash and the emitted values of originator, recipient, amount and nonce. Finally, it transfers MOVE to the recipient address, which has been previously bridged from L1.
5. User is notified on the frontend that their transaction has been completed.

> 6. In case of a rare caase of a reorg on the recipient network that leads to the transaction removal from the user perspective, the transaction can be redone and relayer multisig can handle it.

Reviewer 6: The problem is not a re-org on the target chain, but on the source chain. 
This is explained in the [issue 838](https://github.com/movementlabsxyz/movement/issues/838).

Say if a user is granted k MOVE on L2 on the counterparty via Tx t1, but we discover that the initial transfer on L1 was reverted, how do we undo transactions that have been processed on P2 and depend on t1?
It is not a problem of permission ("the relayer can handle it", that's very hand-wavy), but a problem of undoing tx that could in several different blocks, could impact a multitude of users etc

Reviewer 4: Both reorgs are issues but they are different issues. Just highlighted that in the case of a bug, we guarantee that liveness of Relayer will not impede completion by the maintainer party. In case of a collision a double spend is impossible.

![L2-L1](l2l1.png)

### Key Features

> 1. **Two-Transaction Model**:

Reviewer 6: This is a trusted lock/mint design. Nothing new here.

   - **Initiation**: User sends a transaction to initiate the bridge containing recipient and amount.
   - **Completion**: A relayer or multi-signature group completes the transfer on the counterparty contract with the originator, recipient, amount and noce for hash verification.
   - **No Funds Requirement**: User is not required to have funds on receiving chain and we do not have to build sponsored transactions.
   - **Delivery by Relayer**: Because the relayer delivers the funds, there is no requirement for the user to complete the transaction on receiving chain, therefore massively simplifying the user experience and allowing a smooth onboarding on the network for bridges from L1 to L2. User only has to await for the finalization on the source chain and for the relayer to perform the completion on receiving chain. This is standard practice for every major bridge in the ecosystem.
   - **Less parameters**: Because there is no exchange of secrets between the user and relayer, we have a substantial reduction of logic.

>    - **Only Completable**: Currently we reserve a refunder role to revert transactions. This approach is different where we guarantee delivery of funds through the same party that would guarantee funds being refunded, because bridges can ONLY be completed.

Reviewer 6: I don't understand

> because bridges can ONLY be completed.

Reviewer 4: There is no refund, it can only be completed by the Relayer.

Reviewer 2: changed to 
>    - **Must complete transfers**: Previously we reserve a refunder role to revert transactions. In this approach this is different as we guarantee delivery of funds through the same party that would guarantee funds being refunded, because transfers MUST be completed.

    function initialize(address _moveToken, address _admin, address _relayer, address _maintainer) public initializer {
        require(_moveToken != address(0) && _admin != address(0) && _relayer != address(0), ZeroAddress());
        __Pausable_init();
        moveToken = IERC20(_moveToken);
        _grantRole(DEFAULT_ADMIN_ROLE, _admin);
        _grantRole(RELAYER_ROLE, _relayer);

        // Maintainer is optional
        _grantRole(RELAYER_ROLE, _maintainer);
    }

    function initiateBridgeTransfer(bytes32 recipient, uint256 amount)
        external
        whenNotPaused
        returns (bytes32 bridgeTransferId)
    {
        // Ensure there is a valid amount
        require(amount > 0, ZeroAmount());
        //   _l1l2RateLimit(amount);
        address initiator = msg.sender;

        // Transfer the MOVE tokens from the user to the contract
        if (!moveToken.transferFrom(initiator, address(this), amount)) revert MOVETransferFailed();

        // Generate a unique nonce to prevent replay attacks, and generate a transfer ID
        bridgeTransferId = keccak256(abi.encodePacked(initiator, recipient, amount, ++_nonce));

        // Store the bridge transfer details

>         outgoingBridgeTransfers[bridgeTransferId] = OutgoingBridgeTransfer(initiator, recipient, amount, _nonce);

Reviewer 6: Another thing I don't understand is why we don't use Blobs. There are cheap Ethereum Tx custom made for L2s.

>         outgoingBridgeTransfers[bridgeTransferId] = OutgoingBridgeTransfer(initiator, recipient, amount, _nonce);

Reviewer 4: @franck44 bridgeTransferIds are used for querying the data onchain and completing transactions.
I do advocate for that but in the https://github.com/movementlabsxyz/movement/pull/855 remove struct PR but I wasn't using blobs, just bridgeTransferId verification which still works here.

        emit BridgeTransferCompleted(bridgeTransferId, initiator, recipient, amount, nonce);
    }

    function togglePause() external onlyRole(DEFAULT_ADMIN_ROLE) {
        paused() ? _pause() : _unpause();
    }
}

```

```move
// Based on https://github.com/movementlabsxyz/aptos-core/blob/movement/aptos-move/framework/aptos-framework/sources/atomic_bridge.move  

    /// Module-level resource to store the nonce  
    struct Nonce has key {  
        value: u64,  
    }  

    /// Creates bridge transfer details with validation.  
    ///  
    /// @param initiator The initiating party of the transfer.  
    /// @param recipient The receiving party of the transfer.  
    /// @param amount The amount to be transferred.  
    /// @param nonce The unique nonce for the transfer.  
    /// @return A `BridgeTransferDetails` object.  
    /// @abort If the amount is zero.  
    internal fun create_details<Initiator: store, Recipient: store>(  
        initiator: Initiator,  
        recipient: Recipient,  
        amount: u64,  
        nonce: u64  
    ): BridgeTransferDetails<Initiator, Recipient> {  
        assert!(amount > 0, EZERO_AMOUNT);  

>         BridgeTransferDetails {  

Reviewer 6: there is not transfer status in the details.
How do we prevent someone relaying a relayer Tx with the same parameters?

Reviewer 4: we can't have an status because transaction happens on source then on target. What we do is we check if a value has been placed on that bridgeTransferId, if it has, then it has already been done.

Reviewer 6: I don't see how this is implemented in the code.

Reviewer 4: ```solidity
 require(nonce > 0, InvalidNonce());
        require(idsToIncomingNonces[bridgeTransferId] == 0, CompletedBridgeTransferId());
        // Ensure the bridge transfer ID is valid against the initiator, recipient, amount, and nonce
        require(
            bridgeTransferId == keccak256(abi.encodePacked(initiator, recipient, amount, nonce)),
            InvalidBridgeTransferId()
        );

        // Store the nonce to bridge transfer ID
        idsToIncomingNonces[bridgeTransferId] = nonce;
 ```
 
 I feel like solidity is easier to understand, so the move side should have the same. The only thing missing is the assertion that nonce is over 0

            }  
        );  
    }  

    /// Completes a bridge transfer by the initiator.  
    ///  
    /// @param caller The signer representing the bridge operator.  
    /// @param initiator The initiator's Ethereum address as a vector of bytes.  
    /// @param bridge_transfer_id The unique identifier for the bridge transfer.  
    /// @param recipient The address of the recipient on the Aptos blockchain.  
    /// @param amount The amount of assets to be locked.  
    /// @param nonce The unique nonce for the transfer.  
    
    /// @abort If the caller is not the bridge operator.  
    public entry fun complete_bridge_transfer(  
        caller: &signer,  
        bridge_transfer_id: vector<u8>,
        initiator: vector<u8>,  
        recipient: address,  
        amount: u64,  
        nonce: u64  
    ) {  
        atomic_bridge_configuration::assert_is_caller_operator(caller);  
        let ethereum_address = ethereum::ethereum_address(initiator);
        let details = atomic_bridge_store::create_details(  
            ethereum_address,  
            recipient,  
            amount,  
            nonce  
        );

        // asserts that parameters are correct  

>         assert!(bridge_transfer_id == bridge_transfer_id(&details), 0x1);

Reviewer 6: I don't really understand the check here. 

1. The `initiator` contract on L1 is trusted: the events with the data `bridgeTransferId, initiator, recipient, amount, nonce` are not corrupted and final (assuming we relay only from final blocks). So the `bridgeTransfer` in the event is equal to the `keccak` of the other parameters
2. The relayer is trusted, so it only relays `bridgeTransferId, initiator, recipient, amount, nonce` without tampering with them. 
3. in `complete_bridge_transfer()` (line 387)  we verify that the caller is the operator, which is the relayer (should be).
4. The `create_details` (line 387)  creates a bridge transfer id by hashing the input param (same as keccak) and only checks that this `transferId` has not been completed before.
5. the `assert!` at line 405 should never fail.

All the data `bridgeTransferId, initiator, recipient, amount, nonce` are published on L1 (emitted in an event). 
If I know  `initiator, recipient, amount, nonce`, I can easily compute the `transferId` (line 218, `keccak`).

What added protection does line 405 provide?

Reviewer 2: As per committee agreement, merging will go ahead without addressing this comment. Please consider to raise an issue if this is of concern @franck44

>         atomic_bridge_store::add(bridge_transfer_id, details);

Reviewer 6: This line was supposed to `prepare` the next call (by the user) to complete the bridge transfer.
This part of the algorithm should probably set the status of the transferId to "completed" rather than adding it to the bridge store. 

There should a careful review of what `create_details` and `add` do in the `atomic_bridge_store` to make sure the status of a transfer is updated correctly.

        
        // Mint to recipient  
        atomic_bridge::mint(recipient, amount);  

        event::emit(  
            BridgeTransferCompletedEvent {  
                bridge_transfer_id,  
                initiator,  
                recipient,  
                amount,  
                nonce,  
            },  
        );  
    }  

```

## Verification

1. **Correctness**:
   - Simulate multiple transaction scenarios to ensure robustness.
   - Test edge cases, including relayer downtime and batch processing.

2. **Security Implications**:
   - Conduct audits focused on the simplified design.
   - Implement rate-limiting safeguards and validate parameters in contracts.

3. **Performance Impacts**:
   - Benchmark gas costs and transaction throughput.

4. **Validation Procedures**:
   - Perform an audit and thorough testing, alongside an open invite to the community to verify the bridge.
   - Seek community feedback and incorporate suggestions.

> 

Reviewer 6: What happens if the relayer or network is down for a long time?
What are the trust assumptions?

From the slack discussion I understand the following:
- if the relayer is down or L2 is down, _L2_complete_tx_ will be processed later when the stopped component is up and running again. This is reasonable with the assumption of partial synchrony (eventually the network is synchronous for a period of time and all messages are delivered etc). 
- this requires that:
   - we monitor/health check the relayer, and possibly have a fault-tolerance mechanism in place 
   -  we maintain a _reliable mempool_  on the L2 to make sure tx relayed by the relayer are not lost (e.g. if we reboot the L2 network).
   
It looks like the trust base should  include a reliable  mempool on L2.
When bridging from L2 to L1,  we have to make sure we send our Tx  to an _honest client_ or otherwise it could be censored, delayed etc.

Overall, the only way for a user to get some funds back is to bridge back from L2 to L1.
The duration of a refund is then the time it takes to safely bridge back (minutes, hours, days?)

Reviewer 4: A refund would then be minutes. But I would advocate that we cannot operate it for them, they must process it themselves.

Reviewer 2: > we maintain a reliable mempool on the L2 to make sure tx relayed by the relayer are not lost (e.g. if we reboot the L2 network)

we may be ok without a reliable mempool initially. However, the relayer must check after some time that submitted `complete` transactions have been successful, or otherwise retry them

Reviewer 4: Yes, makes sense.
